### PR TITLE
Feature/crd lifecycle

### PR DESCRIPTION
**Pull request by ycliuhw**
### Checklist

 - ~[ ] Checked if it requires a [pylibjuju](https://github.com/juju/python-libjuju) change?~
 - ~[ ] Added [integration tests](https://github.com/juju/juju/tree/develop/tests) for the PR?~
 - ~[ ] Added or updated [doc.go](https://discourse.jujucharms.com/t/readme-in-packages/451) related to packages changed?~
 - [x] Do comments answer the question of why design decisions were made?

----

## Description of change

Introduce CRD lifecycle. Now charmers can decide when the CRDs get deleted by specifying proper labels.

```json
{
    "juju-resource-lifecycle": "model | persistent"
}
```

1. If no `juju-resource-lifecycle` label set, the CRD gets deleted with the application together.

2. If `juju-resource-lifecycle` sets to `model`, the CRD will not get deleted when the application is removed until the model is destroyed.

3. If `juju-resource-lifecycle` sets to `persistent`, the CRD will never get deleted by Juju even the model is gone.

Driveby fix:

- add validation on labels;
- make the conversion from labels to selector consistent;

## QA steps

deploy a charm has below spec:

```yaml
version: 3
kubernetesResources:
  customResourceDefinitions:
    - name: tfjobs.kubeflow.org
      labels:
        foo: bar  # deleted with the app;
      spec:
        group: kubeflow.org
        scope: Namespaced
        names:
          kind: TFJob
          singular: tfjob
          plural: tfjobs
        versions:
          - name: v1
            served: true
            storage: true
        subresources:
          status: {}
        conversion:
          strategy: None
        validation:
          openAPIV3Schema:
            properties:
              spec:
                properties:
                  tfReplicaSpecs:
                    properties:
                      Worker:
                        properties:
                          replicas:
                            type: integer
                            minimum: 1
                      PS:
                        properties:
                          replicas:
                            type: integer
                            minimum: 1
                      Chief:
                        properties:
                          replicas:
                            type: integer
                            minimum: 1
                            maximum: 1
    - name: tfjob1s.kubeflow.org1
      labels:
        foo: bar
        juju-resource-lifecycle: model  # deleted with the model;
      spec:
        group: kubeflow.org1
        scope: Namespaced
        names:
          kind: TFJob1
          singular: tfjob1
          plural: tfjob1s
        versions:
          - name: v1
            served: true
            storage: true
        subresources:
          status: {}
        validation:
          openAPIV3Schema:
            properties:
              spec:
                properties:
                  tfReplicaSpecs:
                    properties:
                      Worker:
                        properties:
                          replicas:
                            type: integer
                            minimum: 1
                      PS:
                        properties:
                          replicas:
                            type: integer
                            minimum: 1
                      Chief:
                        properties:
                          replicas:
                            type: integer
                            minimum: 1
                            maximum: 1
    - name: tfjob2s.kubeflow.org2
      labels:
        foo: bar
        juju-resource-lifecycle: persistent  # never gets deleted;
      spec:
        group: kubeflow.org2
        scope: Namespaced
        names:
          kind: TFJob2
          singular: tfjob2
          plural: tfjob2s
        versions:
          - name: v1
            served: true
            storage: true
        subresources:
          status: {}
        validation:
          openAPIV3Schema:
            properties:
              spec:
                properties:
                  tfReplicaSpecs:
                    properties:
                     Worker:
                        properties:
                          replicas:
                            type: integer
                            minimum: 1
                      PS:
                        properties:
                          replicas:
                            type: integer
                            minimum: 1
                      Chief:
                        properties:
                          replicas:
                            type: integer
                            minimum: 1
                            maximum: 1

```

```console
$ juju deploy /tmp/charm-builds/mariadb-k8s/ --debug  --resource mysql_image=mariadb -n1

$ mkubectl get crds -o json | jq '.items[] | .metadata | [.name,.labels]'
[
  "tfjob1s.kubeflow.org1",
  {
    "foo": "bar",
    "juju-app": "mariadb-k8s",
    "juju-resource-lifecycle": "model",
    "juju-model": "t1"
  }
]
[
  "tfjob2s.kubeflow.org2",
  {
    "foo": "bar",
    "juju-app": "mariadb-k8s",
    "juju-resource-lifecycle": "persistent",
    "juju-model": "t1"
  }
]
[
  "tfjobs.kubeflow.org",
  {
    "foo": "bar",
    "juju-app": "mariadb-k8s",
    "juju-model": "t1"
  }
]

$ juju remove-application mariadb-k8s -m k1:t1 --destroy-storage --force
removing application mariadb-k8s
- will remove storage database/0

$ mkubectl get crds -o json | jq '.items[] | .metadata | [.name,.labels]'
[
  "tfjob1s.kubeflow.org1",
  {
    "foo": "bar",
    "juju-app": "mariadb-k8s",
    "juju-resource-lifecycle": "model",
    "juju-model": "t1"
  }
]
[
  "tfjob2s.kubeflow.org2",
  {
    "foo": "bar",
    "juju-app": "mariadb-k8s",
    "juju-resource-lifecycle": "persistent",
    "juju-model": "t1"
  }
]

$ juju destroy-model t1 --destroy-storage -y --debug --force

$ mkubectl get crds -o json | jq '.items[] | .metadata | [.name,.labels]'
[
  "tfjob2s.kubeflow.org2",
  {
    "foo": "bar",
    "juju-app": "mariadb-k8s",
    "juju-resource-lifecycle": "persistent",
    "juju-model": "t1"
  }
]

```

## Documentation changes

Yes

## Bug reference

No

### CODE REVIEW

**Comment by ycliuhw** [comment]
[add CRD life cycle](https://trello.com/c/993fESed/1067-add-crd-life-cycle)

File: caas/broker.go [hunk 1]
```diff
@@ -257,8 +257,8 @@ type ServiceGetterSetter interface {
 	// ExposeService sets up external access to the specified service.
 	ExposeService(appName string, resourceTags map[string]string, config application.ConfigAttributes) error
 
-	// UnexposeService removes external access to the specified service.
-	UnexposeService(appName string) error
+	// UnExposeService removes external access to the specified service.
```
Comment (wallyworld) [inline caas/broker.go]: not sure we need this change

File: caas/kubernetes/provider/bootstrap_test.go [hunk 2]
```diff
@@ -689,7 +689,7 @@ $JUJU_TOOLS_DIR/jujud machine --data-dir $JUJU_DATA_DIR --controller-id 0 --log-
 			Return(svcNotProvisioned, nil),
 
 		// below calls are for GetService - 1st address no provisioned yet.
-		s.mockServices.EXPECT().List(v1.ListOptions{LabelSelector: "juju-app==juju-controller-test"}).
+		s.mockServices.EXPECT().List(v1.ListOptions{LabelSelector: "juju-app=juju-controller-test"}).
```
Comment (ycliuhw) [inline caas/kubernetes/provider/bootstrap_test.go]: "==" == "="

File: caas/kubernetes/provider/init.go [hunk 3]
```diff
@@ -12,9 +12,17 @@ import (
...
+var (
+	k8sCloudCheckers             map[string][]k8slabels.Selector
+	jujuPreferredWorkloadStorage map[string]caas.PreferredStorage
+	jujuPreferredOperatorStorage map[string]caas.PreferredStorage
+
+	// labelSelectorGlobalResourcesLifecycleForApplicationRemoval is the label selector for removing global resources for application removal.
+	labelSelectorGlobalResourcesLifecycleForApplicationRemoval k8slabels.Selector
+
+	// labelSelectorGlobalResourcesLifecycleForModelTearDown is the label selector for removing global resources for model teardown.
+	labelSelectorGlobalResourcesLifecycleForModelTearDown k8slabels.Selector
```
Comment (wallyworld) [inline caas/kubernetes/provider/init.go]: Perhaps just 
lifecycleApplicationRemovalSelector
lifecycleModelTeardownSelector

File: caas/kubernetes/provider/admissionregistration_test.go [hunk 4]
```diff
@@ -241,7 +241,7 @@ func (s *K8sBrokerSuite) TestEnsureMutatingWebhookConfigurationsUpdate(c *gc.C)
 		c, cfgs,
 		s.mockMutatingWebhookConfiguration.EXPECT().Create(cfg1).Return(cfg1, s.k8sAlreadyExistsError()),
 		s.mockMutatingWebhookConfiguration.EXPECT().
-			List(metav1.ListOptions{LabelSelector: "juju-app==app-name,juju-model==test"}).
+			List(metav1.ListOptions{LabelSelector: "juju-app=app-name,juju-model=test"}).
```
Comment (wallyworld) [inline caas/kubernetes/provider/admissionregistration_test.go]: Does this have upgrade implications? (are there existing deployments with the "--")

File: caas/kubernetes/provider/k8s.go [hunk 5]
```diff
@@ -73,6 +72,11 @@ const (
 	labelApplicationUUID = "juju-app-uuid"
 	labelModel           = "juju-model"
 
+	// labelGlobalResourceLifeCycleKey defines the label key for lifecycle of the global resources.
+	labelGlobalResourceLifeCycleKey             = "juju-global-resource-lifecycle"
```
Comment (wallyworld) [inline caas/kubernetes/provider/k8s.go]: "juju-resource-lifecycle"
i think is all we needed

File: caas/kubernetes/provider/metadata.go [hunk 6]
```diff
@@ -33,6 +33,30 @@ func newLabelRequirements(rs ...requirementParams) k8slabels.Selector {
...
+}
+
+func labelSetToSelector(labels k8slabels.Set) k8slabels.Selector {
+	return k8slabels.SelectorFromValidatedSet(labels)
+}
+
+func mergeSelectors(selectors ...k8slabels.Selector) k8slabels.Selector {
+	s := k8slabels.NewSelector()
+	for _, v := range selectors {
+		if !v.Empty() {
```
Comment (wallyworld) [inline caas/kubernetes/provider/metadata.go]: maybe invert the condition to reduce indenting

File: caas/kubernetes/provider/specs/types.go [hunk 7]
```diff
@@ -116,6 +117,18 @@ func (cs *k8sContainers) Validate() error {
 	return nil
 }
 
+func validateLabels(labels map[string]string) error {
+	for k, v := range labels {
+		if errs := validation.IsQualifiedName(k); len(errs) != 0 {
+			return errors.NotValidf("invalid label key %q: %s", k, strings.Join(errs, "; "))
+		}
+		if errs := validation.IsValidLabelValue(v); len(errs) != 0 {
+			return errors.NotValidf("invalid label value: %q: at key: %q: %s", v, k, strings.Join(errs, "; "))
```
Comment (wallyworld) [inline caas/kubernetes/provider/specs/types.go]: these error messages will not read correctly

File: caas/kubernetes/provider/specs/v2_test.go [hunk 8]
```diff
@@ -929,6 +932,33 @@ kubernetesResources:
...
+        - http:
+            paths:
+            - path: /testpath
+              backend:
+                serviceName: test
+                servicePort: 80
+`[1:]
+
+	_, err = k8sspecs.ParsePodSpec(specStr)
+	c.Assert(err, gc.ErrorMatches, `invalid label key "/foo": prefix part must be non-empty not valid`)
```
Comment (wallyworld) [inline caas/kubernetes/provider/specs/v2_test.go]: this is an example of the bad error message mentioned above

File: caas/kubernetes/provider/admissionregistration_test.go (hunk 4 above)
Comment (ycliuhw) [reply caas/kubernetes/provider/admissionregistration_test.go]: https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#equality-based-requirement
"=", "=="  represent equality (and are simply synonyms)

**Comment by ycliuhw** [comment]
$$merge$$

**Pull request merged**